src/VisualStudio/Core/Def/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -278,26 +280,39 @@ public static WatcherOperation CombineRange(ImmutableSegmentedList<WatcherOperat
                 }
             }
 
-            return firstOp._kind switch
+            switch (firstOp._kind)
             {
-                Kind.WatchFiles =>
-                    WatchFiles(fileNamesBuilder.ToImmutable(), firstOp._fileChangeFlags, firstOp._sink, tokensBuilder.ToImmutable()),
-                Kind.UnwatchFiles =>
-                    UnwatchFiles(tokensBuilder.ToImmutable()),
-                Kind.UnwatchDirectories =>
-                    UnwatchDirectories(cookiesBuilder.ToList()),
-                _ =>
-                    throw ExceptionUtilities.Unreachable()
-            };
+                case Kind.WatchFiles:
+                    return WatchFiles(fileNamesBuilder.ToImmutable(), firstOp._fileChangeFlags, firstOp._sink, tokensBuilder.ToImmutable());
+                case Kind.UnwatchFiles:
+                    return UnwatchFiles(tokensBuilder.ToImmutable());
+                case Kind.WatchDirectory:
+                    filtersBuilder.RemoveDuplicates();
+                    return WatchDirectory(firstOp._paths[0], filtersBuilder.ToImmutable(), firstOp._sink, cookiesBuilder.ToList());
+                case Kind.UnwatchDirectories:
```

Comment:
There's probably another problem here too: if two watches against a directory are for different filters, and we combine them, what happens on unsubscribe? Do we still keep the other one around?

Reply:
Backed out the watch directory merging, wasn't going to be easy to solve and found a different way to get some watcher reductions.